homeassistant/components/konnected.py

```diff
@@ -0,0 +1,315 @@
+
+REQUIREMENTS = ['konnected==0.1.2']
+
+DOMAIN = 'konnected'
+
+CONF_ACTIVATION = 'activation'
+STATE_LOW = 'low'
+STATE_HIGH = 'high'
+
+PIN_TO_ZONE = {1: 1, 2: 2, 5: 3, 6: 4, 7: 5, 8: 'out', 9: 6}
+ZONE_TO_PIN = {zone: pin for pin, zone in PIN_TO_ZONE.items()}
+
+_BINARY_SENSOR_SCHEMA = vol.All(
+    vol.Schema({
+        vol.Exclusive(CONF_PIN, 's_pin'): vol.Any(*PIN_TO_ZONE),
+        vol.Exclusive(CONF_ZONE, 's_pin'): vol.Any(*ZONE_TO_PIN),
+        vol.Required(CONF_TYPE): DEVICE_CLASSES_SCHEMA,
+        vol.Optional(CONF_NAME): cv.string,
+    }), cv.has_at_least_one_key(CONF_PIN, CONF_ZONE)
+)
+
+_SWITCH_SCHEMA = vol.All(
+    vol.Schema({
+        vol.Exclusive(CONF_PIN, 'a_pin'): vol.Any(*PIN_TO_ZONE),
+        vol.Exclusive(CONF_ZONE, 'a_pin'): vol.Any(*ZONE_TO_PIN),
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Optional(CONF_ACTIVATION, default=STATE_HIGH):
+            vol.All(vol.Lower, vol.Any(STATE_HIGH, STATE_LOW))
+    }), cv.has_at_least_one_key(CONF_PIN, CONF_ZONE)
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema({
+            vol.Required(CONF_ACCESS_TOKEN): cv.string,
+            vol.Required(CONF_DEVICES): [{
+                vol.Required(CONF_ID): cv.string,
+                vol.Optional(CONF_BINARY_SENSORS): vol.All(
+                    cv.ensure_list, [_BINARY_SENSOR_SCHEMA]),
+                vol.Optional(CONF_SWITCHES): vol.All(
+                    cv.ensure_list, [_SWITCH_SCHEMA]),
+            }],
+        }),
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+DEPENDENCIES = ['http', 'discovery']
+
+ENDPOINT_ROOT = '/api/konnected'
+UPDATE_ENDPOINT = (ENDPOINT_ROOT + r'/device/{device_id:[a-zA-Z0-9]+}')
+
+
+async def async_setup(hass, config):
+    """Set up the Konnected platform."""
+    cfg = config.get(DOMAIN)
+    if cfg is None:
+        cfg = {}
+
+    access_token = cfg.get(CONF_ACCESS_TOKEN)
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {CONF_ACCESS_TOKEN: access_token}
+
+    async def async_device_discovered(service, info):
+        """Call when a Konnected device has been discovered."""
+        _LOGGER.debug("Discovered a new Konnected device: %s", info)
+        host = info.get(CONF_HOST)
+        port = info.get(CONF_PORT)
+
+        device = KonnectedDevice(hass, host, port, cfg)
+        await device.async_setup()
+
+    discovery.async_listen(
+        hass,
+        SERVICE_KONNECTED,
+        async_device_discovered)
+
+    hass.http.register_view(KonnectedView(access_token))
+
+    return True
+
+
+class KonnectedDevice(object):
+    """A representation of a single Konnected device."""
+
+    def __init__(self, hass, host, port, config):
+        """Initialize the Konnected device."""
+        self.hass = hass
+        self.host = host
+        self.port = port
+        self.user_config = config
+
+        import konnected
+        self.client = konnected.Client(host, str(port))
+        self.status = self.client.get_status()
```

Comment:
This call is making a request, ie I/O, and we're inside a coroutine here. That's not allowed. Probably change `async_device_discovered` to be non async as well as `KonnectedDevice`. If the interface library is not async, there's usually no point in making the home assistant component async.

Reply:
Hmm, I think this was all originally synchronous but at the suggestion of another reviewer I made these into coroutines: https://github.com/home-assistant/home-assistant/pull/13670#discussion_r183218555
So now I'm thoroughly confused as to the best way to resolve. Seems like either I make everything synchronous or change the `konnected` library to use asyncio http. Which is preferred? Pros/cons of each?